pkg/stress/watcher.go

```diff
@@ -0,0 +1,178 @@
+package stress
+
+import (
+	"context"
+	"math/big"
+	"os/signal"
+	"syscall"
+	"time"
+
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"go.uber.org/zap"
+)
+
+type Watcher struct {
+	ctx             context.Context
+	logger          *zap.Logger
+	ethClient       *ethclient.Client
+	fromBlock       *big.Int
+	watchedContract common.Address
+}
+
+func NewWatcher(
+	ctx context.Context,
+	logger *zap.Logger,
+	rpcURL string,
+	watchedContract common.Address,
+) (*Watcher, error) {
+	ethClient, err := ethclient.Dial(rpcURL)
+	if err != nil {
+		return nil, err
+	}
+
+	block, err := ethClient.BlockByNumber(ctx, nil)
+	if err != nil {
+		ethClient.Close()
+		return nil, err
+	}
```

Comment:
this is probably the parent context? I suspect that neither dial nor blockbynumber have internal timeouts. So this might hang forever.

Reply:
Nice catch, moving the cancellation upwards in the stack.